src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/SqlConnectionX.cs

```diff
@@ -184,13 +184,20 @@ internal Task Close(bool async)
 
             async Task CloseAsync(bool async)
             {
-                Debug.Assert(_connection != null);
-                var internalConnection = _connection;
+                Debug.Assert(_internalConnection != null);
+                Debug.Assert(_dataSource != null);
 
-                //TODO: cancel outstanding operations on connection before close
-                await internalConnection.Close(async).ConfigureAwait(false);
+                var internalConnection = _internalConnection;
 
-                //TODO: remove the internal connection's reference to this wrapper connection
+                if (internalConnection != null)
+                {
+                    //TODO: cancel outstanding operations on connection before close
+                    //TODO: if pooling, reset the connector
+
+                    await internalConnection.Return(async).ConfigureAwait(false);
+                    internalConnection.OwningConnection = null;
```

Comment:
Probably should be done before `Return` as otherwise there's a race condition with other thread renting connection and overwriting `OwningConnection`